### PR TITLE
[AST] Remove ParenType

### PR DESCRIPTION
**Pull request by hamishknight**
Today ParenType is used:

1. As the type of ParenExpr
2. As the payload type of an unlabeled single associated value enum case (and the type of ParenPattern).
3. As the type for an `(X)` TypeRepr

For 1, this leads to some odd behavior, e.g the type of `(5.0 * 5).squareRoot()` is `(Double)`. For 2, we should be checking the arity of the enum case constructor parameters and the presence of ParenPattern respectively. Eventually we ought to consider replacing Paren/TuplePattern with a PatternList node, similar to ArgumentList.

3 is one case where it could be argued that there's some utility in preserving the sugar of the type that the user wrote. However it's really not clear to me that this is particularly desirable since a bunch of diagnostic logic is already stripping ParenTypes. In cases where we care about how the type was written in source, we really ought to be consulting the TypeRepr.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test source compatibility

File: lib/APIDigester/ModuleAnalyzerNodes.cpp [hunk 1]
```diff
@@ -1662,12 +1659,6 @@ SwiftDeclCollector::constructTypeNode(Type T, TypeInitInfo Info) {
 
   SDKNode* Root = SDKNodeInitInfo(Ctx, T, Info).createSDKNode(SDKNodeKind::TypeNominal);
 
-  // Keep paren type as a stand-alone level.
-  if (auto *PT = dyn_cast<ParenType>(T.getPointer())) {
-    Root->addChild(constructTypeNode(PT->getSinglyDesugaredType()));
-    return Root;
-  }
-
```
Comment (hamishknight) [inline lib/APIDigester/ModuleAnalyzerNodes.cpp:1670]: @nkcsgexi Any concerns with removing ParenType from the api-digester? This should only affect when it runs in non-ABI mode

File: lib/Sema/TypeCheckType.cpp [hunk 2]
```diff
@@ -5881,14 +5881,17 @@ NeverNullType TypeResolver::resolveInverseType(InverseTypeRepr *repr,
...
     return ErrorType::get(getASTContext());
 
   // If the inverted type is an existential metatype, unwrap the existential
   // metatype so we can look at the instance type. We'll re-wrap at the end.
-  ExistentialMetatypeType *existentialTy =
-      dyn_cast<ExistentialMetatypeType>(ty.get().getPointer());
+  // Note we don't look through parens.
+  ExistentialMetatypeType *existentialTy = nullptr;
+  if (!constraintRepr->isParenType())
+    existentialTy = dyn_cast<ExistentialMetatypeType>(ty.get().getPointer());
```
Comment (hamishknight) [inline lib/Sema/TypeCheckType.cpp:5894]: @kavon Is it intentional that this doesn't look through parens, e.g we accept `any ~Copyable.Type`, but not `any ~(Copyable.Type)`? For now I've just maintained the current behavior

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
Split out a couple of PRs to hopefully make this a bit easier to land:

- https://github.com/swiftlang/swift/pull/77202
- https://github.com/swiftlang/swift/pull/77203

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test source compatibility

**Review by slavapestov — APPROVED** [review]
Can't wait to see this land!

File: include/swift/Demangling/TypeDecoder.h [hunk 3]
```diff
@@ -1470,16 +1470,6 @@ class TypeDecoder {
 
       return Builder.createDictionaryType(key.getType(), value.getType());
     }
-    case NodeKind::SugaredParen: {
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply include/swift/Demangling/TypeDecoder.h:1473]: Ah okay, does that mean we also need to keep the logic in Demangler.cpp?

[the remark this answers is not in the document]
Comment (hamishknight) [reply include/swift/Demangling/TypeDecoder.h]: Updated with the restored demangling code

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test Windows

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

**Review by xedin — APPROVED** [review]
LGTM! It's slightly unfortunate that we loose the sugar sometimes in the diagnostics but I don't think it's critical.

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please smoke test macOS

**Comment by hamishknight** [comment]
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test macOS

**Comment by hamishknight** [comment]
🚢

**Pull request merged**